core/aws-core/src/main/java/software/amazon/awssdk/awscore/interceptor/FriendlyUnknownHostExceptionInterceptor.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.awscore.interceptor;
+
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.awscore.AwsExecutionAttribute;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.regions.PartitionMetadata;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.regions.RegionMetadata;
+import software.amazon.awssdk.regions.ServiceMetadata;
+import software.amazon.awssdk.regions.ServicePartitionMetadata;
+
+/**
+ * This interceptor will monitor for {@link UnknownHostException}s and provide the customer with additional information they can
+ * use to debug or fix the problem.
+ */
+@SdkInternalApi
+public final class FriendlyUnknownHostExceptionInterceptor implements ExecutionInterceptor {
+    @Override
+    public Throwable modifyException(Context.FailedExecution context, ExecutionAttributes executionAttributes) {
+        // Skip non-UnknownHostExceptions
+        if (!hasCause(context.exception(), UnknownHostException.class)) {
+            return context.exception();
+        }
+
+        StringBuilder error = new StringBuilder();
+        error.append("Received an UnknownHostException when attempting to interact with a service. See cause for the "
+                     + "exact endpoint that is failing to resolve. ");
+
+        Optional<String> globalRegionErrorDetails = getGlobalRegionErrorDetails(executionAttributes);
+
+        if (globalRegionErrorDetails.isPresent()) {
+            error.append(globalRegionErrorDetails.get());
+        } else {
+            error.append("If this is happening on an endpoint that previously worked, there may be a network connectivity "
+                         + "issue or your DNS cache could be storing endpoints for too long.");
+        }
+
+        return SdkClientException.builder().message(error.toString()).cause(context.exception()).build();
```

Comment:
Personally, I stack it when it hits the line limit or when I want to draw more attention to it. I don't stack it when I feel like it's straight-forward and want to draw attention away from it so that the reader's eyes focus more on the interesting stuff.

Reply:
That makes sense to me